pants_plugins/sendwave/pants_node/package.py

```diff
@@ -1,64 +1,141 @@
+
+    This can be useful since, for example, python code copied into a
+    docker container using the pants-docker plugin will already have
+    it's source root stripped. So for the built files to be cleanly
+    layered in to the docker container we should also strip the source
+    roots. There isn't a clean way to do that without using a
+    SourceFiles request, Which I don't think we can create for the new
+    files. Anyway this goes through and finds the root for each item
+    in a snapshot, and removes the source root.
+
+    [0] https://www.pantsbuild.org/v2.9/docs/source-roots
+    """
+    entries = await Get(DigestEntries, Digest, snapshot_to_strip.digest)
+    file_paths, dir_paths = [], []
+    for e in entries:
+        path = PurePath(e.path)
+        if isinstance(e, FileEntry):
+            file_paths.append(path)
+        else:
+            dir_paths.append(path)
+    root_result = await Get(SourceRootsResult,
+                            SourceRootsRequest(files=file_paths,
+                                               dirs=dir_paths))
+    roots = root_result.path_to_root
+    roots_to_entries = {root: [] for (files, root) in roots.items()}
+
+    # collect every output item in the digest by its source root
+    for entry in entries:
+        roots_to_entries[roots[PurePath(entry.path)]].append(entry)
+
+    stripped_digests = []
+    for root, entries in roots_to_entries.items():
+        # create a separate digest for each source root
+        digest = await Get(Digest, CreateDigest(entries))
+        # remove the source root from the digest
+        stripped_digests.append(
+            (await Get(Digest, RemovePrefix(digest, root.path)))
+        )
+    # merge the digests together together
+    return await Get(Digest, MergeDigests(stripped_digests))
+
+
+@dataclass(frozen=True)
+class NodePathRequest:
+    """Empty request to get NodePaths.
+
+    Get requires an argument, but all our configuration is in the
+    NodeSubsystem which is injected separately. So this is just a type
+    marker to tell pants how to give us what we want.
+    """
+
+    pass
+
+
+@dataclass(frozen=True)
+class NPMPath:
+    """Path to npm executable + search path for launched processes."""
+
+    binary_path: str
+    search_paths: Tuple[str]
 
 
 @rule
-async def get_npm_package_files(request: PackageFileRequest) -> Digest:
-    project_root = PurePath(request.package_root)
-    package_json_path = project_root.joinpath("package.json")
-    package_lock = project_root.joinpath("package-lock.json")
-    yarn_lock = project_root.joinpath("yarn.lock")
-    npm_shrinkwrap = project_root.joinpath("npm-shrinkwrap.json")
-
-    rooted_configs = await Get(
-        Digest,
-        PathGlobs(
-            [
-                str(package_json_path),
-                str(package_lock),
-                str(yarn_lock),
-                str(npm_shrinkwrap),
-            ]
+async def get_node_search_paths(request: NodePathRequest,
```

Comment:
This diff is not really helpful since this method does totally different things! Sorry, this just extracts some of the path lookup stuff we were doing before into it's own rule